src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -2,628 +2,317 @@
                 }
             }
         }
 
-        private void NativeCleanup()
-        {
-            // We need to clear the events on websocket as well or stray events
-            // are possible leading to crashes.
-            _innerWebSocket?.RemoveEventListener("close", _onClose);
-            _innerWebSocket?.RemoveEventListener("error", _onError);
-            _innerWebSocket?.RemoveEventListener("open", _onOpen);
-            _innerWebSocket?.RemoveEventListener("message", _onMessage);
-        }
-
         public override void Dispose()
         {
             if (!_disposed)
             {
-                if (_state < (int)InternalState.Aborted) {
-                    _state = (int)InternalState.Disposed;
-                }
                 _disposed = true;
-
-                if (!_cts.IsCancellationRequested)
+                if (_state < WebSocketState.Aborted && _state != WebSocketState.None)
                 {
-                    // registered by the CancellationTokenSource cts in the connect method
-                    _cts.Cancel();
-                    _cts.Dispose();
+                    Abort();
                 }
-
-                _writeBuffer?.Dispose();
-                _receiveMessageQueue.Writer.TryComplete();
-
-                NativeCleanup();
-
+                _state = WebSocketState.Closed;
                 _innerWebSocket?.Dispose();
+                _innerWebSocket = null;
             }
         }
 
-        // This method is registered by the CancellationTokenSource cts in the connect method
-        // and called by Dispose or Abort so that any open websocket connection can be closed.
-        private async void AbortRequest()
+        private async Task ConnectAsyncCore(Uri uri, List<string>? requestedSubProtocols, CancellationToken cancellationToken)
         {
-            switch (State)
+            try
             {
-                case WebSocketState.Open:
-                case WebSocketState.Connecting:
+                object[]? subProtocols = requestedSubProtocols?.ToArray();
+                var onClose = (int code, string reason) =>
+                {
+                    _closeStatus = (WebSocketCloseStatus)code;
+                    _closeStatusDescription = reason;
+                    if (_state == WebSocketState.Connecting || _state == WebSocketState.Open || _state == WebSocketState.CloseSent)
                     {
-                        await CloseAsyncCore(WebSocketCloseStatus.NormalClosure, SR.net_WebSockets_Connection_Aborted, CancellationToken.None).ConfigureAwait(continueOnCapturedContext: true);
-                        // The following code is for those browsers that do not set Close and send an onClose event in certain instances i.e. firefox and safari.
-                        // chrome will send an onClose event and we tear down the websocket there.
-                        if (ReadyStateToDotNetState(_closeStatus) == WebSocketState.CloseSent)
```

Comment:
@campersau do you remember any details about this scenario ? How do it reproduce it and why it matters ?

Reply:
Sadly no, I don't. From the code it looks like in some browsers the `readyState` would still be `2` (CLOSING) when the `onclose` event is called instead of `3` (CLOSED). But I haven't seen that myself only in an old issue of a node.js websocket implementation https://github.com/websockets/ws/issues/1125

https://github.com/dotnet/runtime/pull/47906/commits/011832288eaeac0a4b5f3548427de4a41763ca51